composer/utils/object_store/sftp_object_store.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utility for uploading to and downloading from cloud object stores."""
+import os
+import uuid
+from typing import Optional
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore, ObjectStoreTransientError
+
+__all__ = ['SFTPObjectStore']
+
+
+class SFTPObjectStore(ObjectStore):
+    """Utility for uploading to and downloading to a server via SFTP.
+
+    Args:
+        host (str): The server to connect to
+        port (int): The server port to connect to
+        username (str): The username to authenticate
+        key_file_path (str): The filename of private key
+        cwd (Optional[str]): The directory to navigate to upon creating the SSH connection.
+    """
+
+    def __init__(self, host: str, port: int, username: str, key_file_path: str, cwd: Optional[str] = None):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.key_file_path = key_file_path
+        self.cwd = cwd
+
+        try:
+            from paramiko import SSHClient
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='streaming', conda_package='paramiko') from e
+        self.ssh_client = SSHClient()
+
+        try:
+            self.ssh_client.load_system_host_keys(key_file_path)
+        except IOError:
+            raise Exception('Host keys could not be read from {key_file_path}.')
+
+        self.ssh_client.connect(host, port, username)
+        if cwd is not None:
+            self.ssh_client.exec_command(f'mkdir -p {cwd}')
+            self.ssh_client.exec_command(f'cd {cwd}')
+
+        self.sftp_client = self.ssh_client.open_sftp()
+
+    def close(self):
+        self.ssh_client.close()
+
+    def get_uri(self, object_name: str) -> str:
+        return f'sftp://{self.host}:{self.port}/{object_name}'
+
+    def get_object_size(self, object_name: str) -> int:
+        return self.sftp_client.stat(object_name).st_size
+
+    def upload_object(self, file_path: str, object_name: str):
```

Comment:
The signature should match that of the base class:

```suggestion
    def upload_object(
        self,
        object_name: str,
        filename: Union[str, pathlib.Path],
        callback: Optional[Callable[[int, int], None]] = None,
    ) -> None:
```

Can the ``callback`` parameter be implemented?